app/src/org/commcare/android/database/global/GlobalDatabaseUpgrader.java

```diff
@@ -78,4 +91,40 @@ private boolean upgradeTwoThree(SQLiteDatabase db) {
             db.endTransaction();
         }
     }
+
+    /**
+     * Prior to multiple application seating, the FormsProvider and the InstanceProvider were both
+     * using one global database for all forms/instances. Now that we can have multiple apps
+     * installed at once, we need to have one forms db and one instances db per app. This method
+     * performs the necessary one-time migration from a global db that still exists on a device
+     * to the new per-app system
+     */
+    private boolean upgradeProviderDb(SQLiteDatabase db, ProviderUtils.ProviderType type) {
+        File oldDbFile = CommCareApplication._().getDatabasePath(type.getOldDbName());
+        if (oldDbFile.exists()) {
+            File newDbFile = CommCareApplication._().getDatabasePath(
+                    ProviderUtils.getProviderDbName(type, getInstalledAppRecord(c, db).getApplicationId()));
+            if (!oldDbFile.renameTo(newDbFile)) {
```

Comment:
It seems like it likely should. If we can't rename the file the app won't really work. 

You could do

"Start sqldb transaction for app record.
try to move form provider
try to move instance provider
if both succeed, commit transaciton
if one fails, revert the transaction and restart commcare."

And have the app try again when it restarts.